server/data_common/utils.py

```diff
@@ -0,0 +1,90 @@
+from functools import wraps
+
+from flask import json
+from numpy import float32, integer
+
+from server.common.errors import DriverError
+from enum import Enum
+
+
+class Float32JSONEncoder(json.JSONEncoder):
+    def __init__(self, *args, **kwargs):
+        """
+        NaN/Infinities are illegal in standard JSON.  Python extends JSON with
+        non-standard symbols that most JavaScript JSON parsers do not understand.
+        The `allow_nan` parameter will force Python simplejson to throw an ValueError
+        if it runs into non-finite floating point values which are unsupported by
+        standard JSON.
+        """
+        kwargs["allow_nan"] = False
+        super().__init__(*args, **kwargs)
+
+    def default(self, obj):
+        if isinstance(obj, float32):
+            return float(obj)
+        elif isinstance(obj, integer):
+            return int(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def custom_format_warning(msg, *args, **kwargs):
+    return f"[cellxgene] Warning: {msg} \n"
+
+
+def jsonify_numpy(data):
+    return json.dumps(data, cls=Float32JSONEncoder, allow_nan=False)
+
+
+def requires_data(func):
+    @wraps(func)
+    def wrapped_function(self, *args, **kwargs):
+        if self.data is None:
+            raise DriverError(f"error data must be loaded before you call {func.__name__}")
+        return func(self, *args, **kwargs)
+
+    return wrapped_function
+
+
+class MatrixDataType(Enum):
+    H5AD = "h5ad"
+    TILEDB = "tiledb"
+    UNKNOWN = "unknown"
+
+
+class MatrixDataLoader(object):
+
+    def __init__(self, location):
+        self.location = location
+        self.etype = self.matrix_data_type()
+        self.matrix_type = None
+        if self.etype == MatrixDataType.H5AD:
+            from server.data_scanpy.scanpy_engine import ScanpyEngine
+            self.matrix_type = ScanpyEngine
+        elif self.etype == MatrixDataType.TILEDB:
+            from server.data_tiledb.tiledb_engine import TileDbEngine
+            self.matrix_type = TileDbEngine
+
+    def matrix_data_type(self):
+        if self.location.endswith(".h5ad"):
+            return MatrixDataType.H5AD
+        elif ".cxg" in self.location or ".tdb" in self.location:
+            return MatrixDataType.TILEDB
+        elif self.location.startswith("s3://"):
+            return MatrixDataType.TILEDB
+        else:
+            return MatrixDataType.UNKNOWN
+
+    def pre_checks(self):
```

Comment:
The name was a bit vague.  Suggested rename:  `pre_load_checks` or `pre_load_validation`

Reply:
Done